core/src/test/java/org/pentaho/di/connections/vfs/provider/ConnectionFileNameTest.java

```diff
@@ -0,0 +1,145 @@
+    assertConnectionRoot( "Connection %25 With Percentage In Name", "Connection % With Percentage In Name" );
+
+    assertConnectionRoot( new ConnectionFileName( "connection", null, FileType.FOLDER ), "connection", "connection" );
+    assertConnectionRoot( new ConnectionFileName( "connection", "", FileType.FOLDER ), "connection", "connection" );
+    assertConnectionRoot( new ConnectionFileName( "connection", "/", FileType.FOLDER ), "connection", "connection" );
+  }
+
+  @Test
+  public void testNonRoot() throws Exception {
+    assertNonRoot( "connection", "/file", FileType.FILE );
+    assertNonRoot( "connection", "/folder/file", FileType.FILE );
+
+    // Trims Trailing Path Separator
+    assertComponents(
+      new ConnectionFileName( "connection", "/folder/", FileType.FOLDER ),
+      "connection", "/folder", FileType.FOLDER );
+
+    // Special characters in path
+    String absPath = "/my folder %25 with special Character name &#! <>";
+    ConnectionFileName fileName =
+      new ConnectionFileName( "connection", absPath, FileType.FOLDER );
+    assertComponents( fileName, "connection", absPath, FileType.FOLDER );
+    assertEquals( "/my folder % with special Character name &#! <>", fileName.getPathDecoded() );
+  }
+
+  private static void assertPvsRoot( ConnectionFileName fileName ) throws FileSystemException {
+    assertTrue( fileName.isPvfsRoot() );
+    assertFalse( fileName.isConnectionRoot() );
+    assertEquals( ConnectionFileProvider.SCHEME, fileName.getScheme() );
+    assertNull( fileName.getConnection() );
+    assertNull( fileName.getConnectionDecoded() );
+    assertEquals( FileName.SEPARATOR, fileName.getPath() );
+
+    assertEquals( "pvfs://", fileName.getRootURI() );
+    assertEquals( "pvfs://", fileName.getURI() );
+    assertEquals( "pvfs://", fileName.getFriendlyURI() );
+  }
+
+  private static void assertConnectionRoot( @NonNull String encodedConnectionName ) throws FileSystemException {
+    assertConnectionRoot( encodedConnectionName, encodedConnectionName );
+  }
+
+  private static void assertConnectionRoot( @NonNull String encodedConnectionName,
+                                            @NonNull String decodedConnectionName ) throws FileSystemException {
+    assertConnectionRoot(
+      new ConnectionFileName( encodedConnectionName ),
+      encodedConnectionName,
+      decodedConnectionName );
+  }
+
+  private static void assertConnectionRoot( @NonNull ConnectionFileName fileName,
+                                            @NonNull String encodedConnectionName,
+                                            @NonNull String decodedConnectionName ) throws FileSystemException {
+    assertComponents( fileName, encodedConnectionName, FileName.SEPARATOR, FileType.FOLDER );
+    assertFalse( fileName.isPvfsRoot() );
+    assertTrue( fileName.isConnectionRoot() );
+    assertEquals( decodedConnectionName, fileName.getConnectionDecoded() );
+
+    String uri = "pvfs://" + encodedConnectionName + "/";
```

Comment:
@NJtwentyone

PVFS root connection uris, such as `pvfs://connection/` always end with a slash. This is enforced by Apache VFS for root URIs, and the connection is part of `FileName#getRootURI()` (to which for each distinct value corresponds a distinct `FileSystem` instance).

However, for other contained URIs, historically, the API always removed the trailing `/` of URIs, even for folders...

Would maybe prefer canonicalizing adding the trailing `/` to folder URIs... But didn't want to risk breaking compatiblility at this point...

Did however change the seldom used root URI, from `pvfs:///` to `pvfs://`, given otherwise it would break the expected parent child relationship of URIs. The code tolerates both and normalizes to the second.